ledger/txtail.go

```diff
@@ -45,55 +66,57 @@ type txTail struct {
-	// R + 1 - proto.MaxTxnLife
-	// could not be valid in the next round (R+1), and so are irrelevant.
-	// Thus we load the txids from blocks R+1-maxTxnLife to R, inclusive
-	old := (latest + 1).SubSaturate(basics.Round(proto.MaxTxnLife))
+	var roundData []*txTailRound
+	var roundTailHashes []crypto.Digest
+	var baseRound basics.Round
+	if trackerRound > 0 {
+		err := rdb.Atomic(func(ctx context.Context, tx *sql.Tx) (err error) {
+			roundData, roundTailHashes, baseRound, err = loadTxTail(context.Background(), tx, trackerRound)
+			return err
+		})
+		if err != nil {
+			return err
+		}
+	}
 
-	t.lowWaterMark = latest
+	t.lowWaterMark = trackerRound
 	t.lastValid = make(map[basics.Round]map[transactions.Txid]struct{})
 
-	t.recent = make(map[basics.Round]roundTxMembers)
+	t.recent = make(map[basics.Round]roundLeases)
 
 	// the roundsLastValids is a temporary map used during the execution of
 	// loadFromDisk, allowing us to construct the lastValid maps in their
 	// optimal size. This would ensure that upon startup, we don't preallocate
 	// more memory than we truly need.
 	roundsLastValids := make(map[basics.Round][]transactions.Txid)
 
-	for ; old <= latest; old++ {
-		blk, err := l.Block(old)
-		if err != nil {
-			return err
-		}
+	// allocate with size 0, just so that we can start from fresh.
+	// the roundTailHashes and consensusVersions always has 1 extra element being unused, so preallocate that.
+	roundTailHashes = append([]crypto.Digest{{}}, roundTailHashes...)
+	consensusVersions := make([]protocol.ConsensusVersion, 1)
+	roundTailSerializedData := make([][]byte, 0)
 
-		payset, err := blk.DecodePaysetFlat()
-		if err != nil {
-			return err
-		}
+	for old := baseRound; old <= trackerRound && trackerRound > baseRound; old++ {
+		txTailRound := roundData[0]
+		consensusParams := config.Consensus[txTailRound.ConsensusVersion]
 
-		consensusParams := config.Consensus[blk.CurrentProtocol]
-		t.recent[old] = roundTxMembers{
-			txleases: make(map[ledgercore.Txlease]basics.Round, len(payset)),
+		t.recent[old] = roundLeases{
+			txleases: make(map[ledgercore.Txlease]basics.Round, len(txTailRound.TxnIDs)),
 			proto:    consensusParams,
 		}
 
-		for _, txad := range payset {
-			tx := txad.SignedTxn
-			if consensusParams.SupportTransactionLeases && (tx.Txn.Lease != [32]byte{}) {
-				t.recent[old].txleases[ledgercore.Txlease{Sender: tx.Txn.Sender, Lease: tx.Txn.Lease}] = tx.Txn.LastValid
+		for i := 0; i < len(txTailRound.Leases); i++ {
+			if consensusParams.SupportTransactionLeases && (txTailRound.Leases[i].Lease != [32]byte{}) {
+				t.recent[old].txleases[ledgercore.Txlease{Sender: txTailRound.Leases[i].Sender, Lease: txTailRound.Leases[i].Lease}] = txTailRound.LastValid[txTailRound.Leases[i].TxnIdx]
 			}
-			if tx.Txn.LastValid > t.lowWaterMark {
-				list := roundsLastValids[tx.Txn.LastValid]
+		}
+
+		for i := 0; i < len(txTailRound.LastValid); i++ {
+			if txTailRound.LastValid[i] > t.lowWaterMark {
```

Comment:
Why not include if less than `t.lowWaterMark`?

Reply:
this basically repeats the old logic. If last valid is in past the txn is expired and we are not interested in tracking it.